src/frequenz/sdk/config/_config_managing.py

```diff
@@ -90,32 +85,43 @@ async def _run(self) -> None:
-            # we only react to events related to the configuration file.
-            if not event.path.samefile(self._config_path):
-                continue
-
-            match event.type:
-                case EventType.CREATE:
-                    _logger.info(
-                        "%s: The configuration file %s was created, sending new config...",
-                        self,
-                        self._config_path,
-                    )
-                    await self.send_config()
-                case EventType.MODIFY:
-                    _logger.info(
-                        "%s: The configuration file %s was modified, sending update...",
-                        self,
-                        self._config_path,
-                    )
-                    await self.send_config()
-                case EventType.DELETE:
-                    _logger.info(
-                        "%s: The configuration file %s was deleted, ignoring...",
-                        self,
-                        self._config_path,
-                    )
-                case _:
-                    assert_never(event.type)
+        # FileWatcher can't watch for non-existing files, so we need to watch for the
+        # parent directory instead just in case a configuration file doesn't exist yet
+        # or it is deleted and recreated again.
+        file_watcher = FileWatcher(
+            paths=[self._config_path.parent], event_types=self._event_types
+        )
+
+        try:
+            async for event in file_watcher:
+                # Since we are watching the whole parent directory, we need to make sure
+                # we only react to events related to the configuration file.
+                if not event.path.samefile(self._config_path):
+                    continue
+
+                match event.type:
+                    case EventType.CREATE:
+                        _logger.info(
+                            "%s: The configuration file %s was created, sending new config...",
+                            self,
+                            self._config_path,
+                        )
+                        await self.send_config()
+                    case EventType.MODIFY:
+                        _logger.info(
+                            "%s: The configuration file %s was modified, sending update...",
+                            self,
+                            self._config_path,
+                        )
+                        await self.send_config()
+                    case EventType.DELETE:
+                        _logger.info(
+                            "%s: The configuration file %s was deleted, ignoring...",
+                            self,
+                            self._config_path,
+                        )
+                    case _:
+                        assert_never(event.type)
+        except (Exception, BaseException):
+            del file_watcher
+            raise
```

Comment:
You are manually writing a `finally` :laughing: 

```suggestion
        finally:
            del file_watcher
```

You could also use a [`ExitStack`](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack). Or we could make the `FileWatcher` a context manager (look ma, here it comes `BackgroundService`/`Service` again, if `FileWatcher` were a `Service` it would automatically be a context manager :wink:), but I would leave that for the future as it requires a lot of work (changing `FileWatcher`, releasing etc.) and it should come automatically when we migrate `channels` to use `core` (and `Service`).

Reply:
I really forgot `finally` exists in python, somehow I only thought of try/except/else...sorry :angel: 

> You could also use a [ExitStack](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack). Or we could make the FileWatcher a context manager (look ma, here it comes BackgroundService/Service again, if FileWatcher were a Service it would automatically be a context manager 😉), but I would leave that for the future as it requires a lot of work (changing FileWatcher, releasing etc.) and it should come automatically when we migrate channels to use core (and Service).

Agreed